### PR TITLE
prepare for duty migration in canary checker

### PR DESCRIPTION
**Pull request by adityathebe**
Add new methods and some refactoring to help with https://github.com/flanksource/canary-checker/issues/794

### CODE REVIEW

File: models/components.go [hunk 1]
```diff
@@ -76,17 +67,17 @@ type Component struct {
 	Parents        []string                  `json:"parents,omitempty" gorm:"-"`
 }
 
-func (c *Component) GetStatus() ComponentStatus {
+func (c *Component) GetStatus() types.ComponentStatus {
```
Comment (adityathebe) [inline models/components.go:70]: Though I would just use GetStatus() method from Summary

```go
func (c *Component) GetStatus() types.ComponentStatus {
	return c.Summary.GetStatus()
}
```
but they have different logics. Not sure if it's correct to do it this way.

Comment (adityathebe) [reply models/components.go:70]: Here, I see that component's status is being set using Summary's `GetStatus` method

https://github.com/flanksource/canary-checker/blob/master/pkg/topology/run.go#L399

File: types/component.go [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+	processed bool
+}
+
+func (s *Summary) SetProcessed(val bool) {
+	s.processed = val
+}
+
+func (s Summary) IsProcessed() bool {
+	return s.processed
+}
```
Comment (yashmehrotra) [inline types/component.go:43]: These should be private (lowerCase)

File: types/component.go [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+// GormDBDataType gorm db data type
+func (ResourceSelectors) GormDBDataType(db *gorm.DB, field *schema.Field) string {
+	return JSONGormDBDataType(db.Dialector.Name())
+}
+
+func (rs ResourceSelectors) GormValue(ctx context.Context, db *gorm.DB) clause.Expr {
+	data, _ := json.Marshal(rs)
+	return gorm.Expr("?", string(data))
```
Comment (yashmehrotra) [inline types/component.go:155]: This is also repeating, lets make this generic as well

File: types/config.go [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+	if c.Name != "" {
+		s += "/" + c.Name
+	}
+
+	if len(c.Tags) > 0 {
+		s += " " + fmt.Sprintf("%v", c.Tags)
+	}
+
+	return s
+}
```
Comment (yashmehrotra) [inline types/config.go:42]: Lets just use `%+v` or something like that here ?

File: types/log_selector.go [hunk 5]
```diff
@@ -28,15 +28,7 @@ func (t *LogSelectors) Scan(val any) error {
...
-	switch db.Dialector.Name() {
-	case SqliteType:
-		return JSONType
-	case PostgresType:
-		return JSONBType
-	case SQLServerType:
-		return NVarcharType
-	}
-	return ""
+	return JSONGormDBDataType(db.Dialector.Name())
```
Comment (yashmehrotra) [inline types/log_selector.go:31]: 🚀

File: utils/hash.go [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+		return "", err
+	}
+
+	hash := md5.Sum(data)
+	if err != nil {
+		return "", err
+	}
+
+	return hex.EncodeToString(hash[:]), nil
+}
```
Comment (yashmehrotra) [inline utils/hash.go:22]: Please this to commons/hash, this snippet is used/repeated in other repos too

File: models/components.go [hunk 7]
```diff
@@ -97,40 +89,100 @@ func (component Component) GetAsEnvironment() map[string]interface{} {
...
 
+func (component Component) GetID() string {
+	if component.ID != uuid.Nil {
+		return component.ID.String()
+	}
+	if component.Text != "" {
+		return component.Text
+	}
+	return component.Name
+}
```
Comment (yashmehrotra) [inline models/components.go:134]: Is this function used anywhere ? Ideally it should always return id and throw error on it being nil

File: models/components.go (hunk 1 above)
Comment (yashmehrotra) [reply models/components.go:71]: Hmm, not sure about it. @moshloop Can you chime in here

File: types/config.go (hunk 4 above)
Comment (adityathebe) [reply types/config.go:42]: interestingly, %+v for map makes no difference.


```
fmt.Printf("%v\n", x)
fmt.Printf("%+v\n", x)
fmt.Printf("%#v\n", x)

map[address:global name:flanksource]
map[address:global name:flanksource]
map[string]string{"address":"global", "name":"flanksource"}
```

I think just `%v` looks fine.

File: utils/hash.go (hunk 6 above)
Comment (adityathebe) [reply utils/hash.go:22]: cool. will try to move as much utils from other repos as well to commons.

Comment (adityathebe) [reply utils/hash.go:22]: https://github.com/flanksource/commons/pull/72

File: models/components.go (hunk 7 above)
Comment (adityathebe) [reply models/components.go:133]: I see that it's only used in the `Debug` method. https://github.com/flanksource/duty/blob/4fd7266159f67e96ff1cbd179a5ef9a7daf5cbe0/models/components.go#L207

Comment (yashmehrotra) [reply models/components.go:133]: Ohh, we can delete this method then

File: types/config.go (hunk 4 above)
Comment (yashmehrotra) [reply types/config.go]: Lets just `%+v` or `%v` for the entire String method, since we'll need all the attributes when printing it in logs

**Pull request merged**